azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,138 @@
+import queue
+import logging
+from threading import Event
+
+logger = logging.getLogger(__name__)
+
+
+class MessageQueue(queue.Queue):
+    def __init__(self, on_enable, on_disable):
+        super(MessageQueue, self).__init__()
+        self._enabled = False
+        self._on_enable = on_enable
+        self._on_disable = on_disable
+
+    def enable(self):
+        self._enabled = True
+        self._on_enable()
+
+    def disable(self):
+        self._enabled = False
+        self._on_disable()
+
+    @property
+    def enabled(self):
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, v):
+        if v:
+            self.enable()
+        else:
+            self.disable()
+
+
+class MessageQueueManager(object):
+    def __init__(self, transport):
+        self.c2d_message_queue = MessageQueue(
+            self._c2d_message_queue_on_enable, self._c2d_message_queue_on_disable
+        )  # not yet enabled
+        self.input_message_queues = {}
+        self._transport = transport
+
+    def _input_message_queue_on_enable(self):
+        self._enable_feature("input")
+        # Maybe check if input is enabled first?
+
+    def _c2d_message_queue_on_enable(self):
+        self._enable_feature("c2d")
+
+    def _input_message_queue_on_disable(self):
+        # Disable input messages as a feature if all input queues are disabled
+        if all(not queue.enabled for queue in self.input_message_queues.values()):
+            self._disable_feature("input")
+
+    def _c2d_message_queue_on_disable(self):
+        self._disable_feature("c2d")
+
+    def _enable_feature(self, feature_name):
+        """
+        To enable a specific feature on the internal client.Some of the features that can be enabled
+        are "receiving input messages" and "receiving cloud to device messages"
+        :param feature_name: The specific feature which needs to be enabled. Feature names can be:-
+        "input" , "c2d" etc.
+        "inputs" and on the attribute "on_c2d_message" for "c2d"
+        :raises ValueError if feature_name is not amongst "input" or "c2d"
+        """
+        enable_complete = Event()
+
+        def callback():
+            enable_complete.set()
+
+        if feature_name == "input":
+            self._transport.enable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.enable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
```

Comment:
>ValueError("Feature names can be only among 'input' or 'c2d'") [](start = 18, length = 62)

(opinion) I'm not sure  value (but a lot of potential bugs) in using strings as a way to discriminate which function to call on the transport... I would much rather call something like MessagingManager.c2d.disable() for example. c2d could be an "interface" (or an ABC) that has enable/disable methods

Reply:
This method is private by convention, and is only to be used by us internally, so there's no concern about typos or user error. However the interface has further been adjusted so that this happens at transport level now, getting rid of this logic from client level

---
In reply to: [258258132](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r258258132) [](ancestors = 258258132)